dotnet/src/SemanticKernel/Planning/Plan.cs

```diff
@@ -64,45 +52,141 @@ public Task<Plan> RunNextStepAsync(IKernel kernel, ContextVariables variables, C
     [JsonPropertyName("request_settings")]
-    public CompleteRequestSettings RequestSettings { get; } = new();
+    public CompleteRequestSettings RequestSettings { get; protected set; } = new();
+
+    #endregion ISKFunction implementation
+
+    public Plan(string goal)
+    {
+        this.Description = goal;
+    }
+
+    public Plan(ISKFunction function)
+    {
+        this.SetFunction(function);
+    }
+
+    /// <summary>
+    /// Run the next step of the plan
+    /// </summary>
+    /// <param name="kernel">Kernel instance to use</param>
+    /// <param name="variables">Variables to use</param>
+    /// <param name="cancellationToken">Cancellation token</param>
+    /// <returns>The updated plan</returns>
+    public Task<Plan> RunNextStepAsync(IKernel kernel, ContextVariables variables, CancellationToken cancellationToken = default)
+    {
+        var context = new SKContext(
+            variables,
+            kernel.Memory,
+            kernel.Skills,
+            kernel.Log,
+            cancellationToken);
+        return this.InvokeNextStepAsync(context);
+    }
+
+    #region ISKFunction implementation
 
     /// <inheritdoc/>
     public FunctionView Describe()
     {
-        throw new NotImplementedException();
+        return this.Function?.Describe() ?? new();
     }
 
     /// <inheritdoc/>
-    public Task<SKContext> InvokeAsync(string input, SKContext? context = null, CompleteRequestSettings? settings = null, ILogger? log = null, CancellationToken? cancel = null)
+    public virtual Task<SKContext> InvokeAsync(string input, SKContext? context = null, CompleteRequestSettings? settings = null, ILogger? log = null, CancellationToken? cancel = null)
     {
-        throw new NotImplementedException();
+        if (this.Function is not null)
+        {
+            return this.Function.InvokeAsync(input, context, settings, log, cancel);
+        }
+        else
+        {
+            context ??= new SKContext(new ContextVariables(input), null!, null, log ?? NullLogger.Instance, cancel ?? CancellationToken.None);
+            return this.InvokeAsync(context, settings, log, cancel);
+        }
     }
 
     /// <inheritdoc/>
-    public Task<SKContext> InvokeAsync(SKContext? context = null, CompleteRequestSettings? settings = null, ILogger? log = null, CancellationToken? cancel = null)
+    public virtual async Task<SKContext> InvokeAsync(SKContext? context = null, CompleteRequestSettings? settings = null, ILogger? log = null, CancellationToken? cancel = null)
     {
-        throw new NotImplementedException();
+        if (this.Function is not null)
+        {
+            return await this.Function.InvokeAsync(context, settings, log, cancel);
+        }
+        else
+        {
+            context ??= new SKContext(new ContextVariables(), null!, null, log ?? NullLogger.Instance, cancel ?? CancellationToken.None);
+
+            // loop through steps and execute until completion
+            while (this.Steps.Count > 0)
```

Comment:
Can we improve this?

Reply:
Maybe InvokeNextStepAsync returns bool if there is more to do. IEnumerable but that seems fragile to changes.

Reply:
property of .HasNext or something?

Reply:
If implementors of InvokeNextStepAsync don't decrease, infinite loop.

Reply:
Noted in challenge and adding HasNextStep for now.